File: components/cell-mesh/src/cell_ring.c
// cell_ring.c — fixed-capacity ring of received cells.
//
// Cell bytes are stored canonically (uint8_t[1024]). All header reads go
// through cell_wire.h accessors so there is exactly one representation.

#include "cell_ring.h"

#include <string.h>

void cm_ring_init(cm_ring_t *r) {
    if (!r) return;
    memset(r, 0, sizeof(*r));
}

void cm_ring_push(cm_ring_t *r,
                  const uint8_t cell[CM_CELL_SIZE],
                  const uint8_t peer_mac[6],
                  uint64_t received_at_ms) {
    if (!r || !cell) return;

    uint32_t slot = r->next_index % CM_RING_CAPACITY;
    cm_ring_entry_t *e = &r->entries[slot];

    memcpy(e->cell, cell, CM_CELL_SIZE);
    if (peer_mac) {
        memcpy(e->peer_mac, peer_mac, 6);
    } else {
        memset(e->peer_mac, 0, 6);
    }
    e->received_at_ms = received_at_ms;
    e->occupied       = true;

    r->next_index++;
    r->total_pushed++;
}

size_t cm_ring_visit_newest_first(const cm_ring_t *r,
                                  cm_ring_visit_fn cb,
                                  void *userdata) {
    if (!r || !cb) return 0;

    size_t visited = 0;
    for (size_t i = 0; i < CM_RING_CAPACITY; i++) {
        uint32_t slot = (r->next_index - 1 - (uint32_t)i) % CM_RING_CAPACITY;
        const cm_ring_entry_t *e = &r->entries[slot];
        if (!e->occupied) continue;
        visited++;
        if (!cb(e, userdata)) break;
    }
    return visited;
}

// ── Helper: closure state for cm_ring_count_recent ───────────────────

typedef struct {
    const uint8_t *type_hash;
    uint64_t       now_ms;
    uint32_t       window_ms;
    bool           distinct_peers_only;
    size_t         count;
    uint8_t        seen_peers[CM_RING_CAPACITY][6];
    size_t         seen_peers_n;
} count_recent_ctx_t;

static bool peer_already_seen(count_recent_ctx_t *ctx, const uint8_t mac[6]) {
    for (size_t i = 0; i < ctx->seen_peers_n; i++) {
        if (memcmp(ctx->seen_peers[i], mac, 6) == 0) return true;
    }
    return false;
}

static bool count_recent_cb(const cm_ring_entry_t *e, void *userdata) {
    count_recent_ctx_t *ctx = (count_recent_ctx_t *)userdata;

    // Newest-first iteration: once a cell falls out of window, every
    // further one is older still — stop scanning.
    if (ctx->now_ms < e->received_at_ms) return false;
    uint64_t age_ms = ctx->now_ms - e->received_at_ms;
    if (age_ms > ctx->window_ms) return false;

    // Read type_hash directly from the canonical cell bytes.
    if (memcmp(cm_type_hash(e->cell), ctx->type_hash, 32) != 0) {
        return true; // not a type match; keep scanning
    }

    if (ctx->distinct_peers_only) {
        if (peer_already_seen(ctx, e->peer_mac)) return true;
        if (ctx->seen_peers_n < CM_RING_CAPACITY) {
            memcpy(ctx->seen_peers[ctx->seen_peers_n], e->peer_mac, 6);
            ctx->seen_peers_n++;
        }
    }
    ctx->count++;
    return true;
}

size_t cm_ring_count_recent(const cm_ring_t *r,
                            const uint8_t type_hash[32],
                            uint64_t now_ms,
                            uint32_t window_ms,
                            bool distinct_peers_only) {
    if (!r || !type_hash) return 0;

    count_recent_ctx_t ctx = {
        .type_hash           = type_hash,
        .now_ms              = now_ms,
        .window_ms           = window_ms,
        .distinct_peers_only = distinct_peers_only,
        .count               = 0,
        .seen_peers_n        = 0,
    };
    cm_ring_visit_newest_first(r, count_recent_cb, &ctx);
    return ctx.count;
}

File: components/cell-mesh-zig/src/cell_ring.zig
const std = @import("std");
const wire = @import("cell_wire.zig");

pub const capacity: usize = 16;

pub const RingEntry = extern struct {
    cell: [wire.cell_size]u8,
    peer_mac: [6]u8,
    received_at_ms: u64,
    occupied: bool,
};

pub const Ring = extern struct {
    entries: [capacity]RingEntry,
    next_index: u32,
    total_pushed: u32,
};

pub const VisitFn = *const fn (*const RingEntry, ?*anyopaque) callconv(.c) bool;

pub export fn cm_ring_init(maybe_r: ?*Ring) callconv(.c) void {
    const r = maybe_r orelse return;
    r.* = std.mem.zeroes(Ring);
}

pub export fn cm_ring_push(
    maybe_r: ?*Ring,
    maybe_cell: ?[*]const u8,
    maybe_peer_mac: ?[*]const u8,
    received_at_ms: u64,
) callconv(.c) void {
    const r = maybe_r orelse return;
    const cell = maybe_cell orelse return;

    const slot = r.next_index % capacity;
    var e = &r.entries[slot];

    @memcpy(e.cell[0..], cell[0..wire.cell_size]);
    if (maybe_peer_mac) |peer_mac| {
        @memcpy(e.peer_mac[0..], peer_mac[0..6]);
    } else {
        @memset(e.peer_mac[0..], 0);
    }
    e.received_at_ms = received_at_ms;
    e.occupied = true;

    r.next_index +%= 1;
    r.total_pushed +%= 1;
}

pub export fn cm_ring_visit_newest_first(
    maybe_r: ?*const Ring,
    maybe_cb: ?VisitFn,
    userdata: ?*anyopaque,
) callconv(.c) usize {
    const r = maybe_r orelse return 0;
    const cb = maybe_cb orelse return 0;

    var visited: usize = 0;
    for (0..capacity) |i| {
        const slot = (@as(usize, r.next_index) + capacity - 1 - i) % capacity;
        const e = &r.entries[slot];
        if (!e.occupied) continue;
        visited += 1;
        if (!cb(e, userdata)) break;
    }
    return visited;
}

fn peerAlreadySeen(seen_peers: *[capacity][6]u8, seen_peers_n: usize, mac: *const [6]u8) bool {
    for (seen_peers[0..seen_peers_n]) |seen| {
        if (std.mem.eql(u8, seen[0..], mac[0..])) return true;
    }
    return false;
}

pub export fn cm_ring_count_recent(
    maybe_r: ?*const Ring,
    maybe_type_hash: ?[*]const u8,
    now_ms: u64,
    window_ms: u32,
    distinct_peers_only: bool,
) callconv(.c) usize {
    const r = maybe_r orelse return 0;
    const type_hash = maybe_type_hash orelse return 0;

    var count: usize = 0;
    var seen_peers: [capacity][6]u8 = std.mem.zeroes([capacity][6]u8);
    var seen_peers_n: usize = 0;

    for (0..capacity) |i| {
        const slot = (@as(usize, r.next_index) + capacity - 1 - i) % capacity;
        const e = &r.entries[slot];
        if (!e.occupied) continue;
        if (now_ms < e.received_at_ms) break;
        const age_ms = now_ms - e.received_at_ms;
        if (age_ms > window_ms) break;

        if (!std.mem.eql(u8, e.cell[wire.Off.type_hash..][0..32], type_hash[0..32])) continue;

        if (distinct_peers_only) {
            if (peerAlreadySeen(&seen_peers, seen_peers_n, &e.peer_mac)) continue;
            if (seen_peers_n < capacity) {
                seen_peers[seen_peers_n] = e.peer_mac;
                seen_peers_n += 1;
            }
        }
        count += 1;
    }

    return count;
}
